banyand/property/repair.go

```diff
@@ -0,0 +1,295 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package property
+
+import (
+	"crypto/sha512"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path"
+	"sort"
+	"strconv"
+
+	"github.com/blugelabs/bluge"
+	"github.com/blugelabs/bluge/index"
+	"github.com/cespare/xxhash/v2"
+	"google.golang.org/protobuf/encoding/protojson"
+
+	propertyv1 "github.com/apache/skywalking-banyandb/api/proto/banyandb/property/v1"
+	"github.com/apache/skywalking-banyandb/pkg/convert"
+)
+
+type repair struct {
+	shardPath     string
+	statePath     string
+	treeSlotCount int
+}
+
+func newRepair(shardPath string, treeSlotCount int) *repair {
+	return &repair{
+		shardPath:     shardPath,
+		statePath:     path.Join(shardPath, "state.json"),
+		treeSlotCount: treeSlotCount,
+	}
+}
+
+func (r *repair) buildStatus() (*repairStatus, error) {
+	state, err := r.readState()
+	if err != nil {
+		return nil, fmt.Errorf("reading state failure: %w", err)
+	}
+	indexConfig := index.DefaultConfig(r.shardPath)
+	blugeConf := bluge.DefaultConfig(r.shardPath)
+	items, err := indexConfig.DirectoryFunc().List(index.ItemKindSegment)
+	if err != nil {
+		return nil, fmt.Errorf("reading item kind segment failure: %w", err)
+	}
+	sort.Sort(snapshotIDList(items))
+	// check the snapshot ID have any updated
+	// if no updates, the building Trees should be skipped
+	if state != nil && len(items) != 0 && items[len(items)-1] == state.LastSnapshotID {
+		return state, nil
+	}
+
+	// otherwise, we need to build the Trees
+	tree, err := r.buildingTree(blugeConf, items)
+	if err != nil {
+		return nil, fmt.Errorf("building Trees failure: %w", err)
+	}
+
+	var latestSnapshotID uint64
+	if len(items) > 0 {
+		latestSnapshotID = items[len(items)-1]
+	}
+	// save the Trees to the state
+	state = &repairStatus{
+		LastSnapshotID: latestSnapshotID,
+		Trees:          tree,
+	}
+	stateVal, err := json.Marshal(state)
+	if err != nil {
+		return nil, fmt.Errorf("marshall state failure: %w", err)
+	}
+	err = os.WriteFile(r.statePath, stateVal, 0o600)
+	if err != nil {
+		return nil, fmt.Errorf("writing state file failure: %w", err)
+	}
+	return state, nil
+}
+
+func (r *repair) buildingTree(conf bluge.Config, existingSnapshots []uint64) (map[string]*repairTree, error) {
+	if len(existingSnapshots) == 0 {
+		return make(map[string]*repairTree), nil
+	}
+	reader, err := bluge.OpenReader(conf)
+	if err != nil {
+		return nil, fmt.Errorf("opening index reader failure: %w", err)
+	}
+	totalCount, err := reader.Count()
+	if err != nil {
+		return nil, fmt.Errorf("counting documents failure: %w", err)
+	}
+	treeBuilders := make(map[string]*repairTreeBuilder)
+	for itemID := uint64(0); itemID < totalCount; itemID++ {
```

Comment:
To optimize memory usage, sort the results by `group`, `name`, `entity`, and `timestamp` using the search API. The last entry for the same `group`, `name`, and `entity` should have the largest `timestamp`, indicating the most recent property.

After all properties in a group have been gathered, compute the hashed value for that group. Then, reset the entire builder to prepare for handling the next group. This means you only need to allocate space for a single group at a time.